File: src/DRL/EnvironmentAdapter.cpp
#include "DRL/EnvironmentAdapter.hpp"
#include <algorithm>
#include <cmath>
#include <numeric>

namespace drl {

EnvironmentAdapter::EnvironmentAdapter(int feature_dim) 
    : feature_dim_(feature_dim) {
    initializeDefaults();
    initializeNormalization();
    
    // Define ordered feature names
    feature_names_ = {
        "syscall_count",
        "file_read_count", 
        "file_write_count",
        "file_delete_count",
        "network_connections",
        "bytes_sent",
        "bytes_received", 
        "child_processes",
        "cpu_usage",
        "memory_usage",
        "registry_modification",
        "privilege_escalation_attempt",
        "code_injection_detected",
        "file_io_ratio",
        "network_intensity",
        "process_activity"
    };
}

std::vector<float> EnvironmentAdapter::processTelemetry(const TelemetryData& telemetry) {
    std::vector<float> state(feature_dim_, 0.0f);
    
    // Handle missing fields
    TelemetryData complete_telemetry = handleMissingFields(telemetry);
    
    // Extract basic features
    int idx = 0;
    if (idx < feature_dim_) state[idx++] = normalizeFeature("syscall_count", complete_telemetry.syscall_count);
    if (idx < feature_dim_) state[idx++] = normalizeFeature("file_read_count", complete_telemetry.file_read_count);
    if (idx < feature_dim_) state[idx++] = normalizeFeature("file_write_count", complete_telemetry.file_write_count);
    if (idx < feature_dim_) state[idx++] = normalizeFeature("file_delete_count", complete_telemetry.file_delete_count);
    if (idx < feature_dim_) state[idx++] = normalizeFeature("network_connections", complete_telemetry.network_connections);
    if (idx < feature_dim_) state[idx++] = normalizeFeature("bytes_sent", complete_telemetry.bytes_sent);
    if (idx < feature_dim_) state[idx++] = normalizeFeature("bytes_received", complete_telemetry.bytes_received);
    if (idx < feature_dim_) state[idx++] = normalizeFeature("child_processes", complete_telemetry.child_processes);
    if (idx < feature_dim_) state[idx++] = normalizeFeature("cpu_usage", complete_telemetry.cpu_usage);
    if (idx < feature_dim_) state[idx++] = normalizeFeature("memory_usage", complete_telemetry.memory_usage);
    
    // Boolean features
    if (idx < feature_dim_) state[idx++] = complete_telemetry.registry_modification ? 1.0f : 0.0f;
    if (idx < feature_dim_) state[idx++] = complete_telemetry.privilege_escalation_attempt ? 1.0f : 0.0f;
    if (idx < feature_dim_) state[idx++] = complete_telemetry.code_injection_detected ? 1.0f : 0.0f;
    
    // Derived features
    auto derived = extractDerivedFeatures(complete_telemetry);
    for (size_t i = 0; i < derived.size() && idx < feature_dim_; ++i) {
        state[idx++] = derived[i];
    }
    
    return state;
}

TelemetryData EnvironmentAdapter::handleMissingFields(const TelemetryData& telemetry) {
    TelemetryData complete = telemetry;
    
    // Fill missing numeric fields with defaults
    if (complete.syscall_count < 0) complete.syscall_count = 0;
    if (complete.file_read_count < 0) complete.file_read_count = 0;
    if (complete.file_write_count < 0) complete.file_write_count = 0;
    if (complete.file_delete_count < 0) complete.file_delete_count = 0;
    if (complete.network_connections < 0) complete.network_connections = 0;
    if (complete.bytes_sent < 0) complete.bytes_sent = 0;
    if (complete.bytes_received < 0) complete.bytes_received = 0;
    if (complete.child_processes < 0) complete.child_processes = 0;
    if (complete.cpu_usage < 0.0f || complete.cpu_usage > 100.0f) complete.cpu_usage = 0.0f;
    if (complete.memory_usage < 0.0f) complete.memory_usage = 0.0f;
    
    return complete;
}

void EnvironmentAdapter::updateNormalizationParams(const std::vector<TelemetryData>& telemetry_samples) {
    if (telemetry_samples.empty()) return;
    
    // Compute statistics for each feature
    std::unordered_map<std::string, std::vector<float>> feature_values;
    
    for (const auto& telemetry : telemetry_samples) {
        feature_values["syscall_count"].push_back(telemetry.syscall_count);
        feature_values["file_read_count"].push_back(telemetry.file_read_count);
        feature_values["file_write_count"].push_back(telemetry.file_write_count);
        feature_values["file_delete_count"].push_back(telemetry.file_delete_count);
        feature_values["network_connections"].push_back(telemetry.network_connections);
        feature_values["bytes_sent"].push_back(telemetry.bytes_sent);
        feature_values["bytes_received"].push_back(telemetry.bytes_received);
        feature_values["child_processes"].push_back(telemetry.child_processes);
        feature_values["cpu_usage"].push_back(telemetry.cpu_usage);
        feature_values["memory_usage"].push_back(telemetry.memory_usage);
    }
    
    // Update normalization parameters
    for (const auto& [feature_name, values] : feature_values) {
        if (values.empty()) continue;
        
        auto& params = normalization_params_[feature_name];
        params.min_val = *std::min_element(values.begin(), values.end());
        params.max_val = *std::max_element(values.begin(), values.end());
        
        // Compute mean
        params.mean = std::accumulate(values.begin(), values.end(), 0.0f) / values.size();
        
        // Compute standard deviation
        float variance = 0.0f;
        for (float val : values) {
            variance += (val - params.mean) * (val - params.mean);
        }
        params.std_dev = std::sqrt(variance / values.size());
    }
}

void EnvironmentAdapter::initializeDefaults() {
    default_values_ = {
        {"syscall_count", 0.0f},
        {"file_read_count", 0.0f},
        {"file_write_count", 0.0f},
        {"file_delete_count", 0.0f},
        {"network_connections", 0.0f},
        {"bytes_sent", 0.0f},
        {"bytes_received", 0.0f},
        {"child_processes", 0.0f},
        {"cpu_usage", 0.0f},
        {"memory_usage", 0.0f}
    };
}

void EnvironmentAdapter::initializeNormalization() {
    // Initialize with reasonable defaults based on typical malware behavior
    normalization_params_["syscall_count"] = {0, 10000, 1000, 2000};
    normalization_params_["file_read_count"] = {0, 1000, 50, 100};
    normalization_params_["file_write_count"] = {0, 1000, 50, 100};
    normalization_params_["file_delete_count"] = {0, 100, 5, 10};
    normalization_params_["network_connections"] = {0, 100, 10, 20};
    normalization_params_["bytes_sent"] = {0, 1000000, 10000, 50000};
    normalization_params_["bytes_received"] = {0, 1000000, 10000, 50000};
    normalization_params_["child_processes"] = {0, 50, 2, 5};
    normalization_params_["cpu_usage"] = {0, 100, 30, 20};
    normalization_params_["memory_usage"] = {0, 16000, 500, 1000};
}

float EnvironmentAdapter::normalizeFeature(const std::string& feature_name, float value) {
    auto it = normalization_params_.find(feature_name);
    if (it == normalization_params_.end()) {
        return std::clamp(value, 0.0f, 1.0f);
    }
    
    const auto& params = it->second;
    
    // Min-max normalization
    if (params.max_val > params.min_val) {
        float normalized = (value - params.min_val) / (params.max_val - params.min_val);
        return std::clamp(normalized, 0.0f, 1.0f);
    }
    
    return 0.0f;
}

std::vector<float> EnvironmentAdapter::extractDerivedFeatures(const TelemetryData& telemetry) {
    std::vector<float> derived;
    
    // File I/O ratio
    float total_file_ops = telemetry.file_read_count + telemetry.file_write_count;
    float file_io_ratio = (total_file_ops > 0) ? 
        static_cast<float>(telemetry.file_write_count) / total_file_ops : 0.0f;
    derived.push_back(file_io_ratio);
    
    // Network intensity (normalized to MB)
    float network_intensity = (telemetry.bytes_sent + telemetry.bytes_received) / 1000000.0f;
    derived.push_back(std::clamp(network_intensity, 0.0f, 1.0f));
    
    // Process activity
    float process_activity = std::clamp(telemetry.child_processes / 10.0f, 0.0f, 1.0f);
    derived.push_back(process_activity);
    
    return derived;
}

} // namespace drl

File: python/drl_training/environment_adapter.py
"""
Environment Adapter - Converts raw telemetry to state vectors

This module normalizes telemetry data into fixed-dimension state vectors
suitable for neural network input.
"""

import numpy as np
from typing import Dict, Any, List


class DRLEnvironmentAdapter:
    """
    Adapter for converting raw telemetry data to normalized state vectors.
    
    Handles missing fields, normalization, and maintains consistent feature ordering.
    
    Args:
        feature_dim: Dimension of output state vector
        feature_names: Ordered list of feature names to extract
    """
    
    def __init__(self, feature_dim: int, feature_names: List[str] = None):
        self.feature_dim = feature_dim
        
        # Default feature names if not provided
        if feature_names is None:
            self.feature_names = self._get_default_feature_names()
        else:
            self.feature_names = feature_names
        
        # Default values for missing fields
        self.default_values = self._initialize_defaults()
        
        # Normalization parameters (can be learned from data)
        self.normalization_params = self._initialize_normalization()
    
    def _get_default_feature_names(self) -> List[str]:
        """Get default ordered list of telemetry feature names."""
        return [
            'syscall_count',
            'file_read_count',
            'file_write_count',
            'file_delete_count',
            'network_connections',
            'bytes_sent',
            'bytes_received',
            'child_processes',
            'cpu_usage',
            'memory_usage',
            'registry_modification',
            'privilege_escalation_attempt',
            'code_injection_detected',
            # Additional derived features
            'file_io_ratio',
            'network_intensity',
            'process_activity',
        ]
    
    def _initialize_defaults(self) -> Dict[str, float]:
        """Initialize default values for missing fields."""
        return {
            'syscall_count': 0.0,
            'file_read_count': 0.0,
            'file_write_count': 0.0,
            'file_delete_count': 0.0,
            'network_connections': 0.0,
            'bytes_sent': 0.0,
            'bytes_received': 0.0,
            'child_processes': 0.0,
            'cpu_usage': 0.0,
            'memory_usage': 0.0,
            'registry_modification': 0.0,
            'privilege_escalation_attempt': 0.0,
            'code_injection_detected': 0.0,
            'file_io_ratio': 0.0,
            'network_intensity': 0.0,
            'process_activity': 0.0,
        }
    
    def _initialize_normalization(self) -> Dict[str, Dict[str, float]]:
        """Initialize normalization parameters (min, max, mean, std)."""
        # These should ideally be learned from training data
        # Using reasonable defaults for now
        return {
            'syscall_count': {'min': 0, 'max': 10000, 'mean': 1000, 'std': 2000},
            'file_read_count': {'min': 0, 'max': 1000, 'mean': 50, 'std': 100},
            'file_write_count': {'min': 0, 'max': 1000, 'mean': 50, 'std': 100},
            'file_delete_count': {'min': 0, 'max': 100, 'mean': 5, 'std': 10},
            'network_connections': {'min': 0, 'max': 100, 'mean': 10, 'std': 20},
            'bytes_sent': {'min': 0, 'max': 1000000, 'mean': 10000, 'std': 50000},
            'bytes_received': {'min': 0, 'max': 1000000, 'mean': 10000, 'std': 50000},
            'child_processes': {'min': 0, 'max': 50, 'mean': 2, 'std': 5},
            'cpu_usage': {'min': 0, 'max': 100, 'mean': 30, 'std': 20},
            'memory_usage': {'min': 0, 'max': 16000, 'mean': 500, 'std': 1000},
        }
    
    def process_telemetry(self, telemetry: Dict[str, Any]) -> np.ndarray:
        """
        Convert raw telemetry dictionary to normalized state vector.
        
        Args:
            telemetry: Dictionary containing telemetry data
        
        Returns:
            Normalized state vector of shape (feature_dim,)
        """
        state = np.zeros(self.feature_dim, dtype=np.float32)
        
        # Extract and normalize features
        for i, feature_name in enumerate(self.feature_names[:self.feature_dim]):
            if feature_name in telemetry:
                raw_value = telemetry[feature_name]
                # Convert boolean to float
                if isinstance(raw_value, bool):
                    raw_value = float(raw_value)
                # Normalize
                normalized_value = self._normalize_feature(feature_name, raw_value)
                state[i] = normalized_value
            else:
                # Use default value
                state[i] = self.default_values.get(feature_name, 0.0)
        
        # Compute derived features if needed
        state = self._add_derived_features(state, telemetry)
        
        return state
    
    def _normalize_feature(self, feature_name: str, value: float) -> float:
        """
        Normalize a feature value to [0, 1] range.
        
        Args:
            feature_name: Name of the feature
            value: Raw feature value
        
        Returns:
            Normalized value in [0, 1]
        """
        if feature_name in self.normalization_params:
            params = self.normalization_params[feature_name]
            min_val = params['min']
            max_val = params['max']
            
            # Min-max normalization
            if max_val > min_val:
                normalized = (value - min_val) / (max_val - min_val)
                # Clip to [0, 1]
                normalized = np.clip(normalized, 0.0, 1.0)
                return normalized
        
        # Default: clip to [0, 1]
        return np.clip(value, 0.0, 1.0)
    
    def _add_derived_features(self, state: np.ndarray, 
                             telemetry: Dict[str, Any]) -> np.ndarray:
        """
        Add derived features computed from raw telemetry.
        
        Args:
            state: Current state vector
            telemetry: Raw telemetry dictionary
        
        Returns:
            State vector with derived features
        """
        # Compute file I/O ratio
        file_reads = telemetry.get('file_read_count', 0)
        file_writes = telemetry.get('file_write_count', 0)
        if file_reads + file_writes > 0:
            file_io_ratio = file_writes / (file_reads + file_writes)
        else:
            file_io_ratio = 0.0
        
        # Compute network intensity
        bytes_sent = telemetry.get('bytes_sent', 0)
        bytes_received = telemetry.get('bytes_received', 0)
        network_intensity = (bytes_sent + bytes_received) / 1000000.0  # Normalize to MB
        network_intensity = np.clip(network_intensity, 0.0, 1.0)
        
        # Compute process activity
        child_procs = telemetry.get('child_processes', 0)
        process_activity = np.clip(child_procs / 10.0, 0.0, 1.0)
        
        # Add to state if there's room
        if len(self.feature_names) > len(state):
            # Extend state with derived features
            derived = np.array([file_io_ratio, network_intensity, process_activity], 
                             dtype=np.float32)
            state = np.concatenate([state, derived])
        
        return state[:self.feature_dim]  # Ensure correct dimension
    
    def handle_missing_fields(self, telemetry: Dict[str, Any]) -> Dict[str, Any]:
        """
        Fill in missing fields with default values.
        
        Args:
            telemetry: Telemetry dictionary (possibly incomplete)
        
        Returns:
            Complete telemetry dictionary with defaults filled in
        """
        complete_telemetry = telemetry.copy()
        
        for feature_name in self.feature_names:
            if feature_name not in complete_telemetry:
                complete_telemetry[feature_name] = self.default_values.get(feature_name, 0.0)
        
        return complete_telemetry
